### PR TITLE
Usernames

### PR DESCRIPTION
**Pull request by JPonte**
#### APK
[Download build #8029](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8029/artifact/build/artifact/wire-dev-PR367-8029.apk)
[Download build #8036](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8036/artifact/build/artifact/wire-dev-PR367-8036.apk)
[Download build #8043](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8043/artifact/build/artifact/wire-dev-PR367-8043.apk)
[Download build #8044](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8044/artifact/build/artifact/wire-dev-PR367-8044.apk)
[Download build #8045](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8045/artifact/build/artifact/wire-dev-PR367-8045.apk)
[Download build #8049](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8049/artifact/build/artifact/wire-dev-PR367-8049.apk)
[Download build #8050](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8050/artifact/build/artifact/wire-dev-PR367-8050.apk)
[Download build #8062](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8062/artifact/build/artifact/wire-dev-PR367-8062.apk)
[Download build #8072](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8072/artifact/build/artifact/wire-dev-PR367-8072.apk)
[Download build #8074](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8074/artifact/build/artifact/wire-dev-PR367-8074.apk)
[Download build #8081](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8081/artifact/build/artifact/wire-dev-PR367-8081.apk)
[Download build #8082](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8082/artifact/build/artifact/wire-dev-PR367-8082.apk)
[Download build #8086](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8086/artifact/build/artifact/wire-dev-PR367-8086.apk)
[Download build #8087](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8087/artifact/build/artifact/wire-dev-PR367-8087.apk)
[Download build #8088](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8088/artifact/build/artifact/wire-dev-PR367-8088.apk)
[Download build #8102](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8102/artifact/build/artifact/wire-dev-PR367-8102.apk)
[Download build #8107](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8107/artifact/build/artifact/wire-dev-PR367-8107.apk)

### CODE REVIEW

**Comment by Yorgos-V** [comment]
Some findings after having a look at build 1334-exp

— I did not get the take over screen. I do not have a username.
— I do not see the usernames of other people
— Settings has not been implemented with the latest design. Please advice me when you redo this part.
— We should take the line separators from the profile headers (1:1 or group)
— I should not see common connections info on people I am already connected with

**Comment by xuanzhaopeng** [comment]
1) Doesn't support AB name
2) Doesn't support search

Issues please check here, updated time by time:
https://wearezeta.atlassian.net/issues/?filter=20500

**Comment by shumengye** [comment]
build again

File: app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java [hunk 1]
```diff
@@ -379,6 +398,88 @@ public void onAccentColorHasChanged(Object sender, int color) {
...
+    public void onValidUsernameGenerated(String name, String generatedUsername) {
+        openUsernameAssignTakeOver(name, generatedUsername);
+    }
+
+    @Override
+    public void onUsernameAttemptsExhausted(String name) {
+        openUsernameAssignTakeOver(name, "");
+    }
+
+    private void openUsernameAssignTakeOver(String name, String username) {
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java]: Don't think we need to trigger username screen from here as the opening in the `ConversationListManagerFragment` should cover phones as well

File: app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java [hunk 2]
```diff
@@ -379,6 +398,88 @@ public void onAccentColorHasChanged(Object sender, int color) {
...
+
+    @Override
+    public void onChooseUsernameChosen() {
+        getChildFragmentManager().popBackStackImmediate(FirstTimeAssignUsername.TAG,
+            FragmentManager.POP_BACK_STACK_INCLUSIVE);
+        startActivity(ZetaPreferencesActivity.getUsernameEditPreferencesIntent(getActivity()));
+    }
+
+    @Override
+    public void onKeepUsernameChosen(String username) {
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java]: Same as above, should be covered by `ConversationManagerListFragment`. Also rename method name to `onKeptChosenUsername`

File: app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java [hunk 3]
```diff
@@ -379,6 +398,88 @@ public void onAccentColorHasChanged(Object sender, int color) {
...
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////////////
+    //
+    //  FirstTimeAssignUsername.Container
+    //
+    //////////////////////////////////////////////////////////////////////////////////////////
+
+    @Override
+    public void onChooseUsernameChosen() {
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/pages/main/MainPhoneFragment.java]: Rename method name

**Comment by Yorgos-V** [comment]
I have completed the design review.
There are two issues and, as discussed with @shumengye and @JPonte, they will be taken care of in separate tickets. 
The one is about unifying visually the connection requests screens and the conversation start screens.
The other is about a type style change in the conversation options.

There are two important copy changes that we should fix before releasing.
1. In the search field in the start UI the copy should say: "SEARCH BY NAME OR USERNAME"
2. In the invitations copy, the username should have the @ sign before it. @JPonte will confirm these changes with me.

**Comment by xuanzhaopeng** [comment]
Issue on Tablet:
No takeover (Login in Portrait mode)

**Comment by xuanzhaopeng** [comment]
Issue about emoji on set/change user name:

STR:
1) I have a google keyboard, which support text emoji.
2) I tap text emoji button, and I select anyone

Please check image:
fffuccckk-2016-12-08-1636


Then the App is totally blocked.

**Comment by red-avtovo** [comment]
Issue about showing username on ingoing connection request

STR:
C1 already registered and has no username
C1 sends outgoing connection to C2
C2 sees ingoing connection request (doesn't see username of C1)
C1 adds username
C2 doesn't see username of C1. Even after application restart

**Comment by xuanzhaopeng** [comment]
We will create separated ticket for issues above.

**Comment by vova-lohika** [comment]
Minor issue: oper other user's profile, put wire to background and restore - address book name disappear

**Comment by red-avtovo** [comment]
Changing username removes the avatar

STR:
1)User successfully logged in
2)User go into Settings - Account - Picture
3)User makes a photo and saves it as his avatar
4)User changes it's username
5)User closes the application and opens it again

ER:
User's avatar displays on main screen

AR:
Application loads very long. After loading, user has no avatar

**Pull request merged**